mlflow/tracking/fluent.py

```diff
@@ -401,13 +418,12 @@ def end_run(status: str = RunStatus.to_string(RunStatus.FINISHED)) -> None:
         --
         Active run: None
     """
-    global _active_run_stack, _last_active_run_id
-    if len(_active_run_stack) > 0:
+    if len(_active_run_stack.get()) > 0:
         # Clear out the global existing run environment variable as well.
         MLFLOW_RUN_ID.unset()
-        run = _active_run_stack.pop()
+        run = _active_run_stack.get().pop()
```

Comment:
Is this thread safe? If two concurrent threads both call end_run, what is the return state of the .pop() on the .get() call if the stack experiences the LookupError handling? Won't this throw on line 425?

Reply:
Your answer makes sense :) We should validate this isolation to ensure that references to the active run stack from other modules exhibit no issues. 
A validation where we have 2 threads running outside of context encapsulation of start_run where one terminates long before the other (effectively clearing the stack) and ensuring that the longer running thread still has the ability to call all required fluent APIs (log_params, log_metrics, log_artifacts, tags, etc.) with the other thread's run context stack effectively being None. 
Test suites don't currently exist for this behavior validation because it is known that the current implementation isn't threadsafe. Adding the capability for thread safety requires the addition of some behavioral validations, though :)

Reply:
Yep, makes sense! I added a quick script in the comment below that tests concurrency and fails without this change. I wasn't sure what mlflow's preferred way to write a test for that would be though